File: 프로그래머스/3/43162. 네트워크/네트워크.js
function solution(n, computers) {
    let graph = {};

    // 1. 그래프 만들기
    for (let i = 0; i < n; i++) {
        graph[i] = [];
        for (let j = 0; j < n; j++) {
            if (i !== j && computers[i][j] === 1) {
                graph[i].push(j);
            }
        }
    }

    // 2. 방문 배열 & 네트워크 수 세기
    let visited = Array(n).fill(false);
    let count = 0;

    for (let i = 0; i < n; i++) {
        if (!visited[i]) {
            bfs(i, visited, graph);
            count += 1;
        }
    }

    return count;
}

function bfs(start, visited, graph) {
    let queue = [];
    queue.push(start);
    visited[start] = true;

    while (queue.length > 0) {
        let current = queue.shift();
        for (let next of graph[current]) {
            if (!visited[next]) {
                queue.push(next);
                visited[next] = true;
            }
        }
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] computers) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        
        for (int i = 0; i < n; i ++){
            graph.put(i, new ArrayList<>());
        }
        
        for (int i = 0; i < n; i ++){
            for (int j = 0; j < n; j ++){
                if (i == j) continue;
                if (computers[i][j] == 1){
                    graph.get(i).add(j);
                    graph.get(j).add(i);
                }
            }
        }
        
        int count = 0;
        boolean[] visited = new boolean[n];
        
        for (int i = 0; i < n; i ++){
            if(!visited[i]){
                count += bfs(i, graph, visited);
            }
        }
        return count;
    }
    
    int bfs(int start, Map<Integer, List<Integer>> graph, boolean[] visited){
        Deque<Integer> queue = new ArrayDeque<>();
        queue.offerLast(start);
        visited[start] = true;
        
        while (!queue.isEmpty()){
            int curr = queue.pollFirst();
            
            for (int next: graph.get(curr)){
                if(!visited[next]){
                    visited[next] = true;
                    queue.offerLast(next);
                }
            }
        }
        
        return 1;
    }
}